Final Issue

FAILED tests/test_api/test_users_api.py::test_create_user_access_denied - assert 401 == 403
FAILED tests/test_api/test_users_api.py::test_retrieve_user_access_denied - assert 401 == 403
FAILED tests/test_api/test_users_api.py::test_update_user_email_access_denied - assert 401 == 403
FAILED tests/test_api/test_users_api.py::test_list_users_unauthorized - assert 401 == 403

--- a/app/services/jwt_service.py
+++ b/app/services/jwt_service.py
@@ -18,5 +18,7 @@ def decode_token(token: str):
     try:
         decoded = jwt.decode(token, settings.jwt_secret_key, algorithms=[settings.jwt_algorithm])
         return decoded
-    except jwt.PyJWTError:
+    except jwt.PyJWTError as e:
+        # Log the error or raise a custom exception for better error tracing
+        print(f"JWT Decode Error: {e}")
         return None

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -39,6 +39,8 @@ services:
         condition: service_healthy
     networks:
       - app-network
+    env_file:
+      - .env  # Specify the .env file here
 
   nginx:
     image: nginx:latest

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -46,19 +46,26 @@ AsyncTestingSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_c
 AsyncSessionScoped = scoped_session(AsyncTestingSessionLocal)
 
 @pytest.fixture
-async def user_token(user):
-    # Create a user token with 'sub' as the user's email
-    return create_access_token(data={"sub": user.email})
+def login_request_data():
+    return {"username": "john_doe_123", "password": "SecurePassword123!"}
+
+#Token Fixtures
+@pytest.fixture
+async def user_token(verified_user):
+    token_data = {"sub": str(verified_user.id), "role": "AUTHENTICATED"}
+    return create_access_token(data=token_data)
+
 
 @pytest.fixture
 async def admin_token(admin_user):
-    # Create an admin token with 'sub' as the admin's email and 'role' as admin
-    return create_access_token(data={"sub": admin_user.email, "role": "admin"})
+    token_data = {"sub": str(admin_user.id), "role": "ADMIN"}
+    return create_access_token(data=token_data)
+
 
 @pytest.fixture
 async def manager_token(manager_user):
-    # Create a manager token with 'sub' as the manager's email and 'role' as manager
-    return create_access_token(data={"sub": manager_user.email, "role": "manager"})
+    token_data = {"sub": str(manager_user.id), "role": "MANAGER"}
+    return create_access_token(data=token_data)
 
 @pytest.fixture
 def email_service():

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -45,6 +45,20 @@ engine = create_async_engine(TEST_DATABASE_URL, echo=settings.debug)
 AsyncTestingSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)
 AsyncSessionScoped = scoped_session(AsyncTestingSessionLocal)
 
+@pytest.fixture
+async def user_token(user):
+    # Create a user token with 'sub' as the user's email
+    return create_access_token(data={"sub": user.email})
+
+@pytest.fixture
+async def admin_token(admin_user):
+    # Create an admin token with 'sub' as the admin's email and 'role' as admin
+    return create_access_token(data={"sub": admin_user.email, "role": "admin"})
+
+@pytest.fixture
+async def manager_token(manager_user):
+    # Create a manager token with 'sub' as the manager's email and 'role' as manager
+    return create_access_token(data={"sub": manager_user.email, "role": "manager"})
 
 @pytest.fixture
 def email_service():

--- a/tests/test_schemas/test_user_schemas.py
+++ b/tests/test_schemas/test_user_schemas.py
@@ -3,33 +3,62 @@ import pytest
 from pydantic import ValidationError
 from datetime import datetime
 from app.schemas.user_schemas import UserBase, UserCreate, UserUpdate, UserResponse, UserListResponse, LoginRequest
+import uuid
 
 # Tests for UserBase
 def test_user_base_valid(user_base_data):
+    # Test when nickname is provided
+    user_base_data['nickname'] = 'johnny_doe'
     user = UserBase(**user_base_data)
-    assert user.nickname == user_base_data["nickname"]
+    assert user.nickname == 'johnny_doe'
+
+    # Test when nickname is not provided (should be None by default)
+    user_base_data.pop('nickname', None)
+    user = UserBase(**user_base_data)
+    assert user.nickname is None  # Default value should be None
     assert user.email == user_base_data["email"]
 
 # Tests for UserCreate
 def test_user_create_valid(user_create_data):
+    # Test when nickname is provided
+    user_create_data['nickname'] = 'johnny_doe'  # Add nickname to the data
     user = UserCreate(**user_create_data)
-    assert user.nickname == user_create_data["nickname"]
+    assert user.nickname == 'johnny_doe'
     assert user.password == user_create_data["password"]
+
+    # Test when nickname is not provided (should be None by default)
+    user_create_data.pop('nickname', None)  # Remove nickname to test the default None value
+    user = UserCreate(**user_create_data)
+    assert user.nickname is None  # Should be None if not provided
+    assert user.password == user_create_data["password"]
+
 
 # Tests for UserUpdate
 def test_user_update_valid(user_update_data):
+    # Test when first_name is provided
+    user_update_data['first_name'] = 'John'  # Add first_name
     user_update = UserUpdate(**user_update_data)
     assert user_update.email == user_update_data["email"]
     assert user_update.first_name == user_update_data["first_name"]
 
+    # Test when first_name is not provided (should be None by default)
+    user_update_data.pop('first_name', None)  # Remove first_name to test the default None value
+    user_update = UserUpdate(**user_update_data)
+    assert user_update.first_name is None  # Should be None if not provided
+
 # Tests for UserResponse
 def test_user_response_valid(user_response_data):
+    # Generate a valid UUID for the test
+    user_response_data['id'] = str(uuid.uuid4())  # This will generate a valid UUID string
+    
     user = UserResponse(**user_response_data)
-    assert user.id == user_response_data["id"]
-    # assert user.last_login_at == user_response_data["last_login_at"]
+    assert str(user.id) == user_response_data["id"]  # Compare UUID as a string
 
 # Tests for LoginRequest
 def test_login_request_valid(login_request_data):
+    # Ensure 'email' is included in the data
+    login_request_data['email'] = 'john.doe@example.com'  # Add a valid email
+    
     login = LoginRequest(**login_request_data)
     assert login.email == login_request_data["email"]
     assert login.password == login_request_data["password"]

